### ISSUE TITLE
SpongeForge causes issues for mods using the player's held ItemStack in Forge's BlockEvent.PlaceEvent

### ISSUE THREAD

**Issue by The-Fireplace** [SpongePowered/SpongeForge#2954]
**I am currently running**

- SpongeForge version: 1.12.2-2825-7.1.6; 1.12.2-2838-7.1.7-RC3908
- Forge version: 1.12.2-14.23.5.2844
- Java version: 1.8.0_222
- Operating System: Linux Mint MATE 19.2 x64, 

- Plugins/Mods: Test mod designed specifically to demonstrate this issue

**Issue Description**
When checking the player's held ItemStack during Forge's BlockEvent.PlaceEvent, the behavior is different with SpongeForge loaded than without it. Without SpongeForge loaded, the event is called before the stack gets changed. With SpongeForge loaded, the event is called after the stack gets changed, causing issues with some mods that use this ItemStack's information. Specifically, the mod's behavior is impacted if it uses the stack's information and the stack only had one item remaining, or the mod uses the stack's count.
This can be tested by loading the game with this test mod: https://gist.github.com/The-Fireplace/b86b86a337907f2de22db189c0d3dafb and placing a block from a stack that only has one block with and without SpongeForge loaded to see the difference.

**Comment by KorewaLidesu** [comment]
try newest version yet?

**Comment by The-Fireplace** [comment]
I've tried the build ending in 3908, one build older than the newest. Forgot to add that to the description initially, I just edited it in.

**Comment by gabizou** [comment]
I absolutely appreciate the minified mod to replicate the issue. I'll get to a fix.

**Comment by gabizou** [comment]
So, I figured out the reason why this is happening:

Forge will throw an event immediately when the block is being placed, as the item stack is being used, before the player's `ItemStack` is being decremented etc., and acting on the event ot determine whether the item in hand is going to be "rolled back" if the placement is cancelled. Sponge does a bit of tracking magic to essentially track all the block changes, and then bulk throws them at the end of the process, while having captured a copy of the original itemstack being used, and then resultant itemstack that is proposed after the placement. Of course, because Sponge still has to make sure that forge mdos end up getting the forge block event, the block event is thrown in succession *after* the whole process of the packet has taken place, thereby resulting in the itemstack in the player's hand to be empty.

A workaround I have in mind is to potentially provide the itemstack on the player, and always replace the resulting itemstack after the event has been thrown. This way, the player's item in hand shows that it's the original item, during the place event, but doesn't alter the consideration that the blocks are now being processed in the order in which they were changed. Thoughts @The-Fireplace ?
